app/forms/incomeSources/cease/IncomeSourceEndDateForm.scala

```diff
@@ -84,21 +86,27 @@ class IncomeSourceEndDateForm @Inject()(val dateService: DateService)(implicit v
         "month" -> default(text(), ""),
         "year" -> default(text(), ""))
         .verifying(firstError(
-          checkRequiredFields,
+          checkRequiredFields(incomeSourceType),
           validDate(dateMustNotBeInvalid(incomeSourceType, newIncomeSourceJourney))
         )).transform[LocalDate](
-        {
-          case (day, month, year) =>
-            LocalDate.of(year.toInt, month.toInt, day.toInt)
-        },
-        date => (date.getDayOfMonth.toString, date.getMonthValue.toString, date.getYear.toString)
-      ).verifying(firstError(dateConstraints: _*))
+          {
+            case (day, month, year) =>
+              LocalDate.of(year.toInt, month.toInt, day.toInt)
+          },
+          date => (date.getDayOfMonth.toString, date.getMonthValue.toString, date.getYear.toString)
+        ).verifying(firstError(dateConstraints: _*))
       )(DateFormElement.apply)(DateFormElement.unapply))
   }
 
-  private def checkRequiredFields: Constraint[(String, String, String)] = Constraint("constraints.requiredFields") {
+  private def dateMustBeCompleteKey(incomeSourceType: IncomeSourceType): String = incomeSourceType match {
```

Comment:
nice-to-have: i would move dateMustBeCompleteKey method to below checkRequiredFields method, since dateMustBeCompleteKey is called from checkRequiredFields

Reply:
This function no longer exists from addressing Simon's comment